tpu/templates/tpu_grl_estimator/trainer.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import numpy as np
+import tensorflow as tf
+from tensorflow.contrib import summary
+
+N_CLASSES = 10
+
+
+# ## Gradient Reversal Layer
+#
+# When applied to a tensor this layer is the identity map, but it reverses
+# the sign of the gradient, and optionally multiplies the reversed gradient
+# by a weight.
+#
+# For details, see [Domain-Adversarial Training of Neural Networks](https://arxiv.org/abs/1505.07818).
+#
+
+class GradientReversalLayer(tf.layers.Layer):
+    def __init__(self, weight=1.0):
+        super(GradientReversalLayer, self).__init__()
+        self.weight = weight
+
+    def call(self, input_):
+        @tf.custom_gradient
+        def _call(input_):
+            def reversed_gradient(output_grads):
+                return self.weight * tf.negative(output_grads)
+
+            return input_, reversed_gradient
+        
+        return _call(input_)
+
+
+# ## The model function
+# The network consists of 3 sub-networks:
+#
+# * Feature extractor: extracts internal representation for both the source and target distributions.
+#
+# * Label predictor: predicts label from the extracted features.
+#
+# * Domain classifier: classifies the origin (`source` or `target`) of the extracted features.
+#
+#
+# Both the label predictor and the domain classifier will try to minimize 
+# classification loss, but the gradients backpropagated from the domain
+# classifier to the feature extractor have their signs reversed.
+#
+#
+# This model function also shows how to use `host_call` to output summaries.
+#
+
+def model_fn(features, labels, mode, params):
```

Comment:
can we use params as args using name space: params.gr_weight vs params['gr_weight']

Reply:
I think `params` needs be a dict, as the call to build the estimator sometimes injects/updates additional keys/values to it.